### ISSUE TITLE
[BUG] Unable to invoke models tracked by MLFlow that requires high-dimensional inputs/outputs

### ISSUE THREAD

**Issue by mohammedi-haroune** [mlflow/mlflow#3570]
Thank you for submitting an issue. Please refer to our [issue policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md) for additional information about bug reports. For help with debugging your code, please refer to [Stack Overflow](https://stackoverflow.com/questions/tagged/mlflow).

**Please fill in this bug report template to ensure a timely and thorough response.**

### Willingness to contribute
The MLflow Community encourages bug fix contributions. Would you or another member of your organization be willing to contribute a fix for this bug to the MLflow code base?

- [ ] Yes. I can contribute a fix for this bug independently.
- [x] Yes. I would be willing to contribute a fix for this bug with guidance from the MLflow community.
- [ ] No. I cannot contribute a bug fix at this time.

### System information
- **Have I written custom code (as opposed to using a stock example script provided in MLflow)**:
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 20.10
- **MLflow installed from (source or binary)**: binary
- **MLflow version (run ``mlflow --version``)**: 1.11.0
- **Python version**: 3.7.9 and 3.8.6
- **npm version, if running the dev UI**:
- **Exact command to reproduce**:
Copy the reproduction script below into `reproduction_script.py`
```bash
virtualenv venv --python=python3
source venv/bin/activate
pip install gluoncv==0.8.0 mxnet==1.7.0.post1 mlflow==1.11.0 pandas==1.0.4 numpy==1.18.5 matplotlib==3.3.2
python reproduction_script.py
```
### Describe the problem
**TL;DR;** Models that require a high-dimensional input can't be invoked

The MLFlow `predict` method takes/returns only `pd.DataFrame`s which are [by definition](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html) two dimensional, this makes it impossible to use models that require high-dimensional (>2d) arrays as inputs/outputs, see the reproduction script and the similar open/closed issues 

In fact, many issues were open to allow `predict` to accept/return high-dimensional arrays:
##### Open issues:
* [[BUG] ONNX predict fails for high-dimensional models #3229](https://github.com/mlflow/mlflow/issues/3229)
* [[BUG] Predict N-dimensional data with pyfunc from MLmodel #2830](https://github.com/mlflow/mlflow/issues/2830)
* [[FR] Support tensor inputs to MLflow inference APIs #3159](https://github.com/mlflow/mlflow/issues/3159)

##### Pull requests:
* [Added support for high dimensional onnx models #3307](https://github.com/mlflow/mlflow/pull/3307/)

##### Closed issues:
* [How to pass 2D array as an input to the MLFlow Model #1480](https://github.com/mlflow/mlflow/issues/1480)
* [[FR]How to transfer the multi-dimensional pixel information of the image to the model after the model is deployed as docker？ #1668](https://github.com/mlflow/mlflow/issues/1668)
* [[BUG] Model predict with json input fails if the dataframe has multi-index #2412](https://github.com/mlflow/mlflow/issues/2412)
* [[FR] Mlflow serve - sending an image #2452](https://github.com/mlflow/mlflow/issues/2452)
* [Recommended handling of I/O for pyfunc? #702](https://github.com/mlflow/mlflow/issues/702)
* [TensorFlow wrapper ravels outputs on predict #1066](https://github.com/mlflow/mlflow/issues/1066)

This is clearly something that blocks so many users from using the models that require high-dimensional inputs/outputs logged by MLFlow, I don't understand why MLFlow `predict` method is tightly coupled to `pd.DataFrame`s which clearly can't represent high-dimensional inputs/outputs, why don't we just allow passing `numpy` arrays and return them?  

### Code to reproduce issue
```python
# Run the following pip command to install the dependencies
# ! pip install gluoncv==0.8.0 mxnet==1.7.0.post1 mlflow==1.11.0 pandas==1.0.4 numpy==1.18.5 matplotlib==3.3.2 pyqt5==5.15.1
from gluoncv import model_zoo, data, utils
from matplotlib import pyplot as plt
import pandas as pd
import numpy as np
import mlflow
import mxnet
import mlflow.gluon
import traceback

# Load a pretrained model
net = model_zoo.get_model('yolo3_darknet53_voc', pretrained=True)

# Pre-process an image
im_fname = utils.download('https://raw.githubusercontent.com/zhreshold/' +
                          'mxnet-ssd/master/data/demo/dog.jpg',
                          path='dog.jpg')
x, img = data.transforms.presets.yolo.load_test(im_fname, short=512)
print('Shape of pre-processed image:', x.shape)

# Inference and display
net.hybridize()
class_IDs, scores, bounding_boxs = net(x)

ax = utils.viz.plot_bbox(img, bounding_boxs[0], scores[0],
                         class_IDs[0], class_names=net.classes)


# Log the model with MLFlow
r = mlflow.start_run()
model_name = 'net'
model_path = r.info.artifact_uri + '/' + model_name
mlflow.gluon.log_model(net, model_name)
mlflow.end_run()

# Load the model
m = mlflow.pyfunc.load_model(model_path)


print('-------------------------------------------------------------------------------------------')
print('Using the input as it is')
print('-------------------------------------------------------------------------------------------')
try:
    print(m.predict(x))
except Exception:
    traceback.print_exc()

# Using numpy.ndarray raises AttributeError: 'numpy.ndarray' object has no attribute 'values'
print('-------------------------------------------------------------------------------------------')
print("Using numpy.ndarray raises AttributeError: 'numpy.ndarray' object has no attribute 'values'")
print('-------------------------------------------------------------------------------------------')
try:
    print(m.predict(x.asnumpy()))
except Exception:
    traceback.print_exc()

# Can't construct pd.DataFrame with a >2d shaped array
print('-------------------------------------------------------------------------------------------')
print("Can't construct pd.DataFrame with a >2d shaped array")
print('-------------------------------------------------------------------------------------------')
try:
    print(m.predict(pd.DataFrame(x.asnumpy())))
except Exception:
    traceback.print_exc()


# Flattening the input with np.ravel is not acceptable
print('-------------------------------------------------------------------------------------------')
print("Flattening the input with np.ravel is not acceptable")
print('-------------------------------------------------------------------------------------------')
try:
    print(m.predict(pd.DataFrame(np.ravel(x.asnumpy()))))
except Exception:
    traceback.print_exc()


# Using pd.DataFrame after reshaping also don't work
print('-------------------------------------------------------------------------------------------')
print("Using pd.DataFrame after reshaping also don't work")
print('-------------------------------------------------------------------------------------------')
try:
    x_df = pd.DataFrame(x.reshape(0, -1).asnumpy())
    print(m.predict(x_df))
except Exception:
    traceback.print_exc()
```

### Other info / logs
Here's the output of the reproduction script
```
-------------------------------------------------------------------------------------------
Using the input as it is
-------------------------------------------------------------------------------------------
Traceback (most recent call last):
  File "invoking_mlflow_model_that_requires_multidimensional_input.py", line 45, in <module>
    print(m.predict(x))
  File "/home/mohammedi/Downloads/venv/lib/python3.8/site-packages/mlflow/pyfunc/__init__.py", line 425, in predict
    return self._model_impl.predict(data)
  File "/home/mohammedi/Downloads/venv/lib/python3.8/site-packages/mlflow/gluon.py", line 78, in predict
    ndarray = mx.nd.array(df.values)
AttributeError: 'NDArray' object has no attribute 'values'
-------------------------------------------------------------------------------------------
Using numpy.ndarray raises AttributeError: 'numpy.ndarray' object has no attribute 'values'
-------------------------------------------------------------------------------------------
Traceback (most recent call last):
  File "invoking_mlflow_model_that_requires_multidimensional_input.py", line 54, in <module>
    print(m.predict(x.asnumpy()))
  File "/home/mohammedi/Downloads/venv/lib/python3.8/site-packages/mlflow/pyfunc/__init__.py", line 425, in predict
    return self._model_impl.predict(data)
  File "/home/mohammedi/Downloads/venv/lib/python3.8/site-packages/mlflow/gluon.py", line 78, in predict
    ndarray = mx.nd.array(df.values)
AttributeError: 'numpy.ndarray' object has no attribute 'values'
-------------------------------------------------------------------------------------------
Can't construct pd.DataFrame with a >2d shaped array
-------------------------------------------------------------------------------------------
Traceback (most recent call last):
  File "invoking_mlflow_model_that_requires_multidimensional_input.py", line 63, in <module>
    print(m.predict(pd.DataFrame(x.asnumpy())))
  File "/home/mohammedi/Downloads/venv/lib/python3.8/site-packages/pandas/core/frame.py", line 464, in __init__
    mgr = init_ndarray(data, index, columns, dtype=dtype, copy=copy)
  File "/home/mohammedi/Downloads/venv/lib/python3.8/site-packages/pandas/core/internals/construction.py", line 169, in init_ndarray
    values = prep_ndarray(values, copy=copy)
  File "/home/mohammedi/Downloads/venv/lib/python3.8/site-packages/pandas/core/internals/construction.py", line 295, in prep_ndarray
    raise ValueError("Must pass 2-d input")
ValueError: Must pass 2-d input
-------------------------------------------------------------------------------------------
Flattening the input with np.ravel is not acceptable
-------------------------------------------------------------------------------------------
Traceback (most recent call last):
  File "invoking_mlflow_model_that_requires_multidimensional_input.py", line 73, in <module>
    print(m.predict(pd.DataFrame(np.ravel(x.asnumpy()))))
  File "/home/mohammedi/Downloads/venv/lib/python3.8/site-packages/mlflow/pyfunc/__init__.py", line 425, in predict
    return self._model_impl.predict(data)
  File "/home/mohammedi/Downloads/venv/lib/python3.8/site-packages/mlflow/gluon.py", line 79, in predict
    return pd.DataFrame(self.gluon_model(ndarray).asnumpy())
  File "/home/mohammedi/Downloads/venv/lib/python3.8/site-packages/mxnet/gluon/block.py", line 682, in __call__
    out = self.forward(*args)
  File "/home/mohammedi/Downloads/venv/lib/python3.8/site-packages/mxnet/gluon/block.py", line 1429, in forward
    return self._call_cached_op(x, *args)
  File "/home/mohammedi/Downloads/venv/lib/python3.8/site-packages/mxnet/gluon/block.py", line 1028, in _call_cached_op
    out = self._cached_op(*cargs)
  File "/home/mohammedi/Downloads/venv/lib/python3.8/site-packages/mxnet/_ctypes/ndarray.py", line 148, in __call__
    check_call(_LIB.MXInvokeCachedOpEx(
  File "/home/mohammedi/Downloads/venv/lib/python3.8/site-packages/mxnet/base.py", line 246, in check_call
    raise get_last_ffi_error()
mxnet.base.MXNetError: MXNetError: Error in operator darknetv30_conv0_fwd: [14:32:19] src/operator/nn/convolution.cc:152: Check failed: dshp.ndim() == 4U (2 vs. 4) : Input data should be 4D in batch-num_filter-y-x
-------------------------------------------------------------------------------------------
Using pd.DataFrame after reshaping also don't work
-------------------------------------------------------------------------------------------
Traceback (most recent call last):
  File "invoking_mlflow_model_that_requires_multidimensional_input.py", line 84, in <module>
    print(m.predict(x_df))
  File "/home/mohammedi/Downloads/venv/lib/python3.8/site-packages/mlflow/pyfunc/__init__.py", line 425, in predict
    return self._model_impl.predict(data)
  File "/home/mohammedi/Downloads/venv/lib/python3.8/site-packages/mlflow/gluon.py", line 79, in predict
    return pd.DataFrame(self.gluon_model(ndarray).asnumpy())
  File "/home/mohammedi/Downloads/venv/lib/python3.8/site-packages/mxnet/gluon/block.py", line 682, in __call__
    out = self.forward(*args)
  File "/home/mohammedi/Downloads/venv/lib/python3.8/site-packages/mxnet/gluon/block.py", line 1429, in forward
    return self._call_cached_op(x, *args)
  File "/home/mohammedi/Downloads/venv/lib/python3.8/site-packages/mxnet/gluon/block.py", line 1028, in _call_cached_op
    out = self._cached_op(*cargs)
  File "/home/mohammedi/Downloads/venv/lib/python3.8/site-packages/mxnet/_ctypes/ndarray.py", line 148, in __call__
    check_call(_LIB.MXInvokeCachedOpEx(
  File "/home/mohammedi/Downloads/venv/lib/python3.8/site-packages/mxnet/base.py", line 246, in check_call
    raise get_last_ffi_error()
mxnet.base.MXNetError: MXNetError: Error in operator darknetv30_conv0_fwd: [14:32:19] src/operator/nn/convolution.cc:152: Check failed: dshp.ndim() == 4U (2 vs. 4) : Input data should be 4D in batch-num_filter-y-x
```

### What component(s), interfaces, languages, and integrations does this bug affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

**Comment by apcragg** [comment]
Any update on this? I don't understand how any ML software can not support tensors as inputs. Do the authors expect people to only use MLflow for toy problems such as the Iris classifier?

**Comment by vlad-tokarev** [comment]
So, I think it should be implemented as yet another generic flavor. Like pyfunc flavor but with tensor signature.

Kubeflow project has an initiative about [cross-platform prediction API for model servers ](https://github.com/kubeflow/kfserving/blob/master/docs/predict-api/v2/required_api.md)

This API is already supported by Triton ML Server and some other mature solutions. It looks very generic and extensible, for example, it supports N input tensors, not just one. That is also a limitation for some frameworks such as PyTorch

Of course, this is a specification of REST/gRPC API – not Python API.

But it could be mapped on python API.
For example, look at how Python ML Server [declares signature for a Sklearn model API](https://github.com/SeldonIO/MLServer/blob/3ddf2714653c526f8938c8070e5723cce8aedcab/mlserver/models/sklearn.py#L42)

What do you think about such an approach?  Thanks you!

**Comment by jhadjar** [comment]
@apcragg @vlad-tokarev We're going with custom flavors to support that for [iko.ai](https://iko.ai) but outside of the MLFlow main tree for several reasons:

- We automatically parametrize notebooks for our users, and then automatically track models/params/metrics, and allow them to deploy the resulting model or build its Docker image by clicking a button without them having to do anything. i.e: it is against our design ideal to have users bear the cognitive load of writing adapters or worrying about tracking.

- We cannot afford to wait for our hypothetical changes to be merged, as there are other pull requests for similar issues that are hanging for months, and issues about this problem tend to be closed with comments by the MLflow team not seeing how this is a problem. In our experience, most non toy models are high-dimensional.

- Forking MLflow to support that will hinder us because we'll have to maintain that fork and very likely have to deal with conflicts pulling in upstream changes.

Therefore, we're developing flavors to support that in a modular way, which is something we're comfortable with, given that this is one of our core design principles for everything we do on [iko.ai](https://iko.ai).

**Comment by patryk-oleniuk** [comment]
Waiting for this issue to close as well. In my opinion I should be able to pass multi-dimentional arrays as a json to the REST api, just like this: 
 Body, for example `np.array` of `shape=(1, 2, 3, 1)` :
```json
[[ [0.0] , [0.1], [0.2], [0.3] ],  [ [1.0] , [1.1], [1.2], [1.3] ]]
```
Could help with dev if needed

I dislike that it has to be passed as pandas dfs very much, since many DNNs don't operate with pandas or JSONs.

**Comment by jhadjar** [comment]
@patryk-oleniuk @mateiz replied to a tweet[1]. Quoting:

> We're currently working on adding multi-dimensional tensors as an input data type, but people have been handling that in two ways:
> - For images, audio, etc it's often better to pass in a binary file (that way it's also compressed).
> - You can also pass a 1D vector and reshape it.


[1]: https://twitter.com/matei_zaharia/status/1327342800766046214

**Comment by patryk-oleniuk** [comment]
Nice, thanks @jhadjar. Of course you can do that kind of workaround, but you need wrapper like here: [flower_claffifier_example](https://github.com/mlflow/mlflow/tree/master/examples/flower_classifier)

It is unacceptable for me to write so much code just to workaround this issue, there should be a better way. Currently it makes MLflow model serving not really usable for most image-processing DNN projects. 

Very happy to know you guys are working on it and will definitely use it when it's ready.
